### PR TITLE
Dragging - Prevent multiple units carrying the same object

### PR DESCRIPTION
**Pull request by LinkIsGrim**
**When merged this pull request will:**
- Title.

### IMPORTANT

- If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

**Comment by johnb432** [comment]
Did you get any errors from https://github.com/acemod/ACE3/blob/a5da6cc83d554002debca4064989382c6466cfc8/addons/common/functions/fnc_claim.sqf#L25

Alternatively we could make use of [`EFUNC(common,claimSafeServer)`](https://github.com/acemod/ACE3/blob/master/addons/common/functions/fnc_claimSafeServer.sqf). It would require a bit of refactoring, but it would be the most robust imo.

**Comment by johnb432** [comment]
> Alternatively we could make use of [`EFUNC(common,claimSafeServer)`](https://github.com/acemod/ACE3/blob/master/addons/common/functions/fnc_claimSafeServer.sqf). It would require a bit of refactoring, but it would be the most robust imo.

https://github.com/acemod/ACE3/tree/no-multi-user-carry

Completely untested, but I imagine it *should* be fairly straight forward.

**Comment by LinkIsGrim** [comment]
> Did you get any errors from https://github.com/acemod/ACE3/blob/a5da6cc83d554002debca4064989382c6466cfc8/addons/common/functions/fnc_claim.sqf#L25
> 
> ?
> 
> ---
> 
> Alternatively we could make use of [`EFUNC(common,claimSafeServer)`](https://github.com/acemod/ACE3/blob/master/addons/common/functions/fnc_claimSafeServer.sqf). It would require a bit of refactoring, but it would be the most robust imo.
> 

I wrote this while tabbed out of the game mid-session. Completely untested.

File: addons/dragging/functions/fnc_setDraggable.sqf [hunk 1]
```diff
@@ -56,8 +56,25 @@ GVAR(initializedClasses) = _initializedClasses;
...
 private _icon = [QPATHTOF(UI\icons\box_drag.paa), QPATHTOF(UI\icons\person_drag.paa)] select (_object isKindOf "CAManBase");
 
-private _dragAction = [QGVAR(drag), LLSTRING(Drag), _icon, {[_player, _target] call FUNC(startDrag)}, {[_player, _target] call FUNC(canDrag)}] call EFUNC(interact_menu,createAction);
-private _dropAction = [QGVAR(drop), LLSTRING(Drop), "", {[_player, _target] call FUNC(dropObject)}, {[_player, _target] call FUNC(canDrop)}] call EFUNC(interact_menu,createAction);
+private _dragAction = [
+    QGVAR(drag),
+    LLSTRING(Drag),
+    _icon,
+    {
+        [QEGVAR(common,claimSafe), [_player, _target, true, QGVAR(startDrag)]] call CBA_fnc_serverEvent;
```
Comment (LinkIsGrim) [inline addons/dragging/functions/fnc_setDraggable.sqf:64]: Maybe this should be moved into a function (`startDragLocal`?)?

**Comment by LinkIsGrim** [comment]
LGTM otherwise (john's code). Tested in dedicated MP.

File: addons/dragging/functions/fnc_startDrag.sqf
Comment (LinkIsGrim) [inline addons/dragging/functions/fnc_startDrag.sqf:1]: Moved without changes to `fnc_startDragLocal.sqf`

File: addons/dragging/functions/fnc_startCarry.sqf
Comment (LinkIsGrim) [inline addons/dragging/functions/fnc_startCarry.sqf:1]: moved without changes to `fnc_startCarryLocal.sqf`

File: addons/dragging/functions/fnc_startDrag.sqf
Comment (johnb432) [reply addons/dragging/functions/fnc_startDrag.sqf:1]: I like this - keeps this backward compatible with 3rd party mods (I know it's not public, but still).

I'm guessing you have already tested this new version?

File: addons/dragging/functions/fnc_setCarryable.sqf [hunk 2]
```diff
@@ -56,8 +56,25 @@ GVAR(initializedClasses_carry) = _initializedClasses;
 
 private _icon = [QPATHTOF(UI\icons\box_carry.paa), QPATHTOF(UI\icons\person_carry.paa)] select (_object isKindOf "CAManBase");
 
-private _carryAction = [QGVAR(carry), LLSTRING(Carry), _icon, {[_player, _target] call FUNC(startCarry)}, {[_player, _target] call FUNC(canCarry)}] call EFUNC(interact_menu,createAction);
-private _dropAction = [QGVAR(drop_carry), LLSTRING(Drop), "", {[_player, _target] call FUNC(dropObject_carry)}, {[_player, _target] call FUNC(canDrop_carry)}] call EFUNC(interact_menu,createAction);
+private _carryAction = [
```
Comment (johnb432) [inline addons/dragging/functions/fnc_setCarryable.sqf:59]: Since nothing has changed here, we can revert these changes, unless the new formatting is preferred.

File: addons/dragging/functions/fnc_setDraggable.sqf [hunk 3]
```diff
@@ -56,8 +56,25 @@ GVAR(initializedClasses) = _initializedClasses;
 
 private _icon = [QPATHTOF(UI\icons\box_drag.paa), QPATHTOF(UI\icons\person_drag.paa)] select (_object isKindOf "CAManBase");
 
-private _dragAction = [QGVAR(drag), LLSTRING(Drag), _icon, {[_player, _target] call FUNC(startDrag)}, {[_player, _target] call FUNC(canDrag)}] call EFUNC(interact_menu,createAction);
-private _dropAction = [QGVAR(drop), LLSTRING(Drop), "", {[_player, _target] call FUNC(dropObject)}, {[_player, _target] call FUNC(canDrop)}] call EFUNC(interact_menu,createAction);
+private _dragAction = [
```
Comment (johnb432) [inline addons/dragging/functions/fnc_setDraggable.sqf:59]: Same.

File: addons/dragging/functions/fnc_startCarry.sqf [hunk 4]
```diff
@@ -11,82 +11,11 @@
...
-
-// Disable collisions by setting the PhysX mass to almost zero
-private _mass = getMass _target;
 
-if (_mass > 1) then {
-    _target setVariable [QGVAR(originalMass), _mass, true];
-    [QEGVAR(common,setMass), [_target, 1e-12]] call CBA_fnc_globalEvent; // Force global sync
-};
+// Try to claim the object
+[QEGVAR(common,claimSafe), [_unit, _targer, true, QGVAR(startCarry)]] call CBA_fnc_serverEvent;
```
Comment (johnb432) [inline addons/dragging/functions/fnc_startCarry.sqf:21]:
```suggestion
[QEGVAR(common,claimSafe), [_unit, _target, true, QGVAR(startCarry)]] call CBA_fnc_serverEvent;
```

File: addons/dragging/functions/fnc_startDrag.sqf [hunk 5]
```diff
@@ -1,104 +1,21 @@
...
-
-// Disable collisions by setting the physx mass to almost zero
-private _mass = getMass _target;
 
-if (_mass > 1) then {
-    _target setVariable [QGVAR(originalMass), _mass, true];
-    [QEGVAR(common,setMass), [_target, 1e-12]] call CBA_fnc_globalEvent; // Force global sync
-};
+// Try to claim the object
+[QEGVAR(common,claimSafe), [_unit, _targer, true, QGVAR(startDrag)]] call CBA_fnc_serverEvent;
```
Comment (johnb432) [inline addons/dragging/functions/fnc_startDrag.sqf:21]:
```suggestion
[QEGVAR(common,claimSafe), [_unit, _target, true, QGVAR(startDrag)]] call CBA_fnc_serverEvent;
```

**Review by johnb432 — CHANGES_REQUESTED** [review]

File: addons/dragging/functions/fnc_setCarryable.sqf (hunk 2 above)
Comment (LinkIsGrim) [reply addons/dragging/functions/fnc_setCarryable.sqf:59]: Meh. It's more readable this way IMO

**Review by johnb432 — APPROVED** [review]

File: addons/dragging/functions/fnc_startDrag.sqf
Comment (johnb432) [reply addons/dragging/functions/fnc_startDrag.sqf:1]: I've tested on my end, LGTM.
